rust/rsc/src/rsc/blob.rs

```diff
@@ -1,71 +1,96 @@
-        }
-        let name = field.name().unwrap().to_string();
-        let Ok(data) = field.bytes().await else {
-            break false;
-        };
+async fn stream_to_file<S, E>(
+    parent: &str,
+    path: &str,
+    stream: S,
+) -> Result<(), (StatusCode, String)>
+where
+    S: Stream<Item = Result<Bytes, E>>,
+    E: Into<BoxError>,
+{
+    if !path_is_valid(path) {
+        return Err((StatusCode::BAD_REQUEST, "Invalid path".to_owned()));
+    }
 
-        if name == "hash" {
-            let key = String::from_utf8(data.to_ascii_lowercase()).unwrap();
-            hash = Some(key);
-        }
+    async {
+        // Convert the stream into an `AsyncRead`.
+        let body_with_io_error =
+            stream.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err));
+        let body_reader = StreamReader::new(body_with_io_error);
+        futures::pin_mut!(body_reader);
 
-        if name == "file" {
-            contents = Some(data);
-        }
-    };
+        // Create the file. `File` implements `AsyncWrite`.
+        let path = std::path::Path::new(parent).join(path);
+        let mut file = BufWriter::new(File::create(path).await?);
+
+        // Copy the body into the file.
+        tokio::io::copy(&mut body_reader, &mut file).await?;
 
-    if !exhausted_without_err {
-        return hyper::StatusCode::BAD_REQUEST;
+        Ok::<_, std::io::Error>(())
     }
+    .await
+    .map_err(|err| (StatusCode::INTERNAL_SERVER_ERROR, err.to_string()))
+}
 
-    let Some(hash) = hash else {
-        return hyper::StatusCode::BAD_REQUEST;
-    };
+#[tracing::instrument]
+pub async fn get_upload_url(server_addr: String) -> Json<GetUploadUrlResponse> {
+    let url = server_addr + "/blob";
+    Json(GetUploadUrlResponse { url })
+}
 
-    let Some(contents) = contents else {
-        return hyper::StatusCode::BAD_REQUEST;
-    };
+#[tracing::instrument]
+pub async fn create_blob(
+    mut multipart: Multipart,
+    _conn: Arc<DatabaseConnection>,
+    config: Arc<RSCConfig>,
+) -> (StatusCode, String) {
+    while let Ok(Some(field)) = multipart.next_field().await {
+        let file_name = if let Some(file_name) = field.name() {
+            file_name.to_owned()
+        } else {
+            tracing::warn!("Skipping part upload due to missing key.");
+            continue;
+        };
 
-    let Some(local_store) = config.local_store.clone() else {
-        return hyper::StatusCode::INTERNAL_SERVER_ERROR;
-    };
+        let Some(ref local_store) = config.local_store else {
+            return (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "Invalid configuration".into(),
+            );
+        };
 
-    let dest = local_store + &hash;
-    let Ok(_) = std::fs::write(dest, contents) else {
-        return hyper::StatusCode::INTERNAL_SERVER_ERROR;
-    };
+        // TODO: instead of writing the user provided field name to disk we should
+        // generate a random temporary name.
```

Comment:
```
use rand_core::{RngCore, OsRng};
use data_encoding::BASE64;

let mut key = [0u8; 16];
OsRng.fill_bytes(&mut key);
let dst = local_store + "/" + BASE64.encode(&key);
```

The online hashing part would still need to be done as well of course but I'm happy delaying that. I just don't think we should commit code that uses a user specified location as the destination.

Reply:
I think my priorities are

1. Do not commit code with security flaws in it at all
2. Try to make each PR a "functional" member (happy to slip on this one)
3. Be incremental to keep PR size down

I'm very happy to give ground on 2 to make 1 hold

Reply:
re: 1

Out of self preservation I'll call out that it doesn't actually have a security flaw. It has a _potential_ for a security flaw that is avoided by erroring out if the user gives bad input.

That said, there isn't any reason to just use a random name from the start so I'll go ahead and do that